chore(dot/parachain): improve Run method of subsystems

axaysagathiya opened the pull request:
## Changes
- I have updated the Run method of the subsystem interface to pass only necessary arguments.
- removed context, context cancel func, wait group and OverseerToSubSystem channel from the state struct of all the subsystems(example: candidateBacking{}, AvailabilityStoreSubsystem{}).
- improved Run method of some subsystems.
- removed return type from RegisterSubsystem method of overseer.
- because of these changes, some tests were failing, so I fixed those tests.


## Tests

<!-- Detail how to run relevant tests to the changes -->

```sh
go test ./dot/parachain/...
```

## Issues
Closes #4110 
<!-- Write the issue number(s), for example: #123 -->

axaysagathiya committed:
pass context cancel to Run method of subsystem

axaysagathiya committed:
lint

axaysagathiya committed:
Merge branch 'feat/parachain' of github.com:ChainSafe/gossamer into axay/chore/parachain/use-context-properly

axaysagathiya committed:
clean up

P1sar commented on dot/parachain/candidate-validation/candidate_validation.go:
```diff
@@ -43,9 +43,19 @@ func NewCandidateValidation(overseerChan chan<- any) *CandidateValidation {
 
 // Run starts the CandidateValidation subsystem
-func (cv *CandidateValidation) Run(context.Context, chan any, chan any) {
-	cv.wg.Add(1)
```
I guess we then should also remove wg from CandidateValidaton? It is not used anymore

P1sar commented on dot/parachain/candidate-validation/candidate_validation.go:
```diff
@@ -43,9 +43,19 @@ func NewCandidateValidation(overseerChan chan<- any) *CandidateValidation {
+			if err != nil {
+				logger.Error(err.Error())
+			}
+		case <-cv.stopChan:
```
I do not see any place where this stop channel is getting initiated or used (written to).

axaysagathiya replied:
I have just moved the code. I have not written any logic other than passing the `ctx` and `cancel` to the Run method.
But still, I will look into this and try to improve

axaysagathiya replied:
Removed `stopChan` and used `ctx.Done()` instead

P1sar commented on dot/parachain/candidate-validation/candidate_validation.go:
```diff
@@ -43,9 +43,19 @@ func NewCandidateValidation(overseerChan chan<- any) *CandidateValidation {
-func (cv *CandidateValidation) Run(context.Context, chan any, chan any) {
-	cv.wg.Add(1)
-	go cv.processMessages(&cv.wg)
+func (cv *CandidateValidation) Run(context.Context, context.CancelFunc, chan any, chan any) {
```
Why non of this arguments are used?

axaysagathiya replied:
I am planning to fix this. 

Run is a method of subsystem interface that we implement in all the subsystems.
As far as I have noticed, 
- the last argument is not used in any of the subsystems. 
- The second last argument is used by some subsystems only.
- context cancel doesn't seem to be required

So I am going to take some time and fix everything.

P1sar commented on dot/parachain/candidate-validation/candidate_validation.go:
```diff
@@ -72,46 +82,37 @@ func (cv *CandidateValidation) Stop() {
+func (cv *CandidateValidation) processMessages(msg any) error {
+	switch msg := msg.(type) {
+	case ValidateFromChainState:
+		// TODO: implement functionality to handle ValidateFromChainState, see issue #3919
```
Maybe would be nice to add panic with this TODO message?

P1sar commented on dot/parachain/candidate-validation/candidate_validation.go:
```diff
@@ -72,46 +82,37 @@ func (cv *CandidateValidation) Stop() {
 		}
+
+	case PreCheck:
+		// TODO: implement functionality to handle PreCheck, see issue #3921
```
Maybe would be nice to add panic with this TODO message?

axaysagathiya replied:
updated

EclesioMeloJunior commented on dot/parachain/availability-store/availability_store.go:
```diff
@@ -420,8 +415,12 @@ func branchesFromChunks(chunks [][]byte) (branches, error) {
-func (av *AvailabilityStoreSubsystem) Run(ctx context.Context, OverseerToSubsystem chan any,
-	SubsystemToOverseer chan any) {
+func (av *AvailabilityStoreSubsystem) Run(
+	ctx context.Context, cancel context.CancelFunc,
```
Its common to only passes the context to the `Run` method but not the cancelation function, like:

```go
ctx, cancel := context.WithCancel(context.Background())
av.Run(ctx)
```

Does not make sense to give to the av the possibility to cancel the context itself owns, also since this context is used by `av.processMessages` I would suggest to pass the context to it then you don't need to keep it as a field in `AvailabilityStoreSubsystem` , for example:

```go
func (av *AvailabilityStoreSubsystem) Run(ctx context.Context) {
    ...
    go av.processMessages(ctx)
}

func (av *AvailabilityStoreSubsystem) processMessages(ctx context.Context) {
    ...
    for {
       select {
       case <-ctx.Done():
             ...
       }
    }
}
```

axaysagathiya replied:
Updated the code 
- to only pass `ctx` to the Run method, not the cancel function.
- Used the same context in Run instead of storing context in the struct(like candidateBacking, AvailabilityStoreSubsystem, etc.)
- and added the `ctx.Done()` case in the Run method of all the subsystems which didn't have it already.

axaysagathiya committed:
Merge branch 'feat/parachain' into axay/chore/parachain/use-context-properly

axaysagathiya committed:
remove stopCh and Wg from candidate validation

axaysagathiya committed:
remove cancel from Run method of subsystem

axaysagathiya committed:
go mod tidy

axaysagathiya committed:
remove OverseerToSubsystem field from all subsystems

- need to remove return type from overseer RegisterSubsystem method
- some tests are failing in this commit

axaysagathiya committed:
fix test before removing return type of overseer RegisterSubsystem

P1sar commented on dot/parachain/availability-store/availability_store.go:
```diff
@@ -420,89 +410,87 @@ func branchesFromChunks(chunks [][]byte) (branches, error) {
-	av.wg.Add(1)
-	go av.processMessages()
+func (av *AvailabilityStoreSubsystem) Run(
+	ctx context.Context, OverseerToSubsystem chan any, SubsystemToOverseer chan any,
```
since OverseerToSubsystem will be only readed from and SubsystemToOverseer will be only written to, would it make sense to pass them as directed channels arguments?

axaysagathiya replied:
updated

axaysagathiya committed:
remove return type of RegisterSubsystem method

- modified TestHandleBlockEvents to not use the returned value from
  RegisterSubsystem method.
- still we have multiple errors in 'validator_side_test.go' file because
  they are using retuned value from RegisterSubsystem method.

axaysagathiya committed:
Merge branch 'feat/parachain'

axaysagathiya committed:
remove SubSystemToOverseer channel from Run method

axaysagathiya committed:
improve Run method of network bridge

axaysagathiya committed:
lint

axaysagathiya committed:
make overseerToSubSystem the receive only channel

axaysagathiya committed:
fix tests

axaysagathiya committed:
mockgen

axaysagathiya committed:
panic for todo

EclesioMeloJunior commented on dot/parachain/candidate-validation/candidate_validation.go:
```diff
@@ -73,67 +78,56 @@ func (*CandidateValidation) ProcessBlockFinalizedSignal(parachaintypes.BlockFina
+	switch msg := msg.(type) {
+	case ValidateFromChainState:
+		runtimeInstance, err := cv.BlockState.GetRuntime(msg.CandidateReceipt.Descriptor.RelayParent)
+		if err != nil {
```
should you return inside this condition?

axaysagathiya replied:
Yes, Updated it

EclesioMeloJunior reviewed (changes requested):
LGTM, just a small comment

axaysagathiya committed:
address review

kishansagathiya commented on dot/parachain/availability-store/availability_store.go:
```diff
@@ -420,89 +409,85 @@ func branchesFromChunks(chunks [][]byte) (branches, error) {
-				if err != nil {
-					logger.Errorf("failed to process block finalized signal: %w", err)
-				}
+func (av *AvailabilityStoreSubsystem) processMessages(msg any) {
```
this has now become processMessage :)
apply this in all subsystems when you have made this change

axaysagathiya replied:
updated

kishansagathiya reviewed (approved):
:tada:

axaysagathiya committed:
rename processMessages method to processMessage
